blockchain/chain/src/store/chain_store.rs

```diff
@@ -225,6 +250,26 @@ where
     Ok((bls_msgs, secp_msgs))
 }
 
+/// Constructs and returns a full tipset if messages from storage exists - non self version
+pub fn fill_tipsets<DB>(db: &DB, ts: Tipset) -> Result<FullTipset, Error>
+where
+    DB: BlockStore,
+{
+    let mut blocks: Vec<Block> = Vec::with_capacity(ts.blocks().len());
+
+    for header in ts.into_blocks() {
+        let (bls_messages, secp_messages) = block_messages(db, &header)?;
+        blocks.push(Block {
+            header,
+            bls_messages,
+            secp_messages,
+        });
+    }
+
+    // the given tipset has already been verified, so this cannot fail
+    Ok(FullTipset::new(blocks).unwrap())
+}
```

Comment:
can you have the function you copied this from call this so we don't have duplication here?